### ISSUE TITLE
`__torch_function__` does not get call for torch.Tensor's `__getitem__` syntax

### ISSUE THREAD

**Issue by zdevito** [pytorch/pytorch#46277]
## 🐛 Bug

If you write `tensor_val[my_custom_type_with_torch_function]`, the `__torch_function__` call is not invoked on `my_custom_type_with_torch_function` even though it is defined. This works for other operators like `__add__`
`
## To Reproduce

```
import torch
class MyTensor:
    def __torch_function__(self, orig_method, types, args=None, kwargs=None):
        args = args if args else ()
        kwargs = kwargs if kwargs else {}
        print("CALLED ", orig_method)

a = torch.rand(3, 4)
index = MyTensor()

a + index # works: CALLED <method 'add' of 'torch._C._TensorBase' objects>
a[index] # fails: IndexError: only integers, slices (`:`), ellipsis (`...`), None and long or byte Variables are valid indices (got MyTensor)
```

## Expected behavior

`__getitem__` works the same way as `__add__`

## Environment

```
Collecting environment information...
PyTorch version: 1.8.0a0+f7b9046
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] numpy==1.17.0
[pip3] torch==1.8.0a0
[pip3] torch-struct==0.4
[pip3] torchfile==0.1.0
[pip3] torchtext==0.8.0a0+db31b5d
[pip3] torchvision==0.8.0a0+fc49090
[pip3] torchviz==0.0.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.1                      217  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.1.0            py37h23d657b_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.17.0                   pypi_0    pypi
[conda] torch                     1.8.0a0                   dev_0    <develop>
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchtext                 0.8.0a0+db31b5d           dev_0    <develop>
[conda] torchvision               0.8.0a0+fc49090           dev_0    <develop>
[conda] torchviz                  0.0.1                    pypi_0    pypi
```

## Additional context

This is causing issues while symbolically tracing models with torch.fx, e.g.:
https://github.com/pytorch/benchmark/blob/5768afa082e8931424f803edfca1ac4a627b6143/torchbenchmark/models/Super_SloMo/slomo_model.py#L325

**Comment by zdevito** [comment]
@ezyang

**Comment by ezyang** [comment]
cc @hameerabbasi @rgommers

**Comment by hameerabbasi** [comment]
This works somewhat but not completely. The error message is misleading, this happens because you're trying to index by a `float`, which would error in any case.

```
>>> import torch
>>> a = torch.rand((3, 4))
>>> class MyTensor(torch.Tensor):
...     pass
... 
>>> b = a.as_subclass(MyTensor)
>>> index = torch.randint(3, ())
>>> type(a[index])
<class 'torch.Tensor'>
>>> type(b[index])
<class '__main__.MyTensor'>
>>> indexb = index.as_subclass(MyTensor)
>>> type(a[indexb]) # Borked
<class 'torch.Tensor'>
>>> type(b[indexb]) # Works
<class '__main__.MyTensor'>
```

**Comment by zdevito** [comment]
In the case where we run into this, `MyTensor` is not a Tensor (nor does it contain one, it is a symbolic representation of a Tensor), so trying to use `as_subclass` isn't really viable. When I looked at the code for handling `__getitem__` in C++ it looks like it is only checking `self` for `__torch_function__`, but is ignoring the potential `__torch_function__` on the index variable.

**Comment by hameerabbasi** [comment]
> In the case where we run into this, `MyTensor` is not a Tensor (nor does it contain one, it is a symbolic representation of a Tensor), so trying to use `as_subclass` isn't really viable. When I looked at the code for handling `__getitem__` in C++ it looks like it is only checking `self` for `__torch_function__`, but is ignoring the potential `__torch_function__` on the index variable.

This is correct. Although it's more complicated than that, if it's a `tuple`/`list` being passed into `__getitem__`, we need to check _every input_ for `__torch_function__` if it isn't an exact `Tensor` or Python type. I seem to recall I couldn't figure out a way to do that at the time without dropping performance, or maybe it was overly complicated.

**Comment by hameerabbasi** [comment]
@ezyang What would be the best thing to do here? See the above comment. Do we want this fixed at the cost of performance?

**Comment by ezyang** [comment]
I feel like it should still be feasible to do this test without too much worse performance, because we're going to be parsing the tuple no matter what (similar to how we have the append overloaded args thing).

**Comment by hameerabbasi** [comment]
I'm working on a fix for this in #49324.
